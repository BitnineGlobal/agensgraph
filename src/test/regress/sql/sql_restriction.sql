--
-- SQL Restriction with Graph objects
--

-- prepare
CREATE GRAPH g;
CREATE ROLE tmp;

--
-- ALTER SCHEMA
--

ALTER SCHEMA g RENAME TO s;
ALTER SCHEMA g OWNER TO tmp;

--
-- CREATE TABLE
--

CREATE TABLE g.t (i int);
CREATE TABLE t (i int) INHERITS (g.ag_vertex);

--
-- ALTER TABLE
--

GRANT ALL ON DATABASE regression TO tmp;
SET ROLE tmp;

CREATE GRAPH t;
SET graph_path = t;

CREATE VLABEL v;

ALTER TABLE t.v ADD COLUMN tmp int;

RESET ROLE;

SET graph_path = g;

CREATE VLABEL v;
ALTER TABLE g.v RENAME TO e;

--
-- TRUNCATE TABLE
--
TRUNCATE TABLE g.v;

--
-- TRIGGER
--

CREATE TRIGGER tt AFTER INSERT ON g.v
FOR EACH STATEMENT EXECUTE PROCEDURE ff();

--
-- RULE
--

CREATE VLABEL v2;
CREATE RULE rr AS ON INSERT TO g.v DO INSTEAD
	INSERT INTO g.v2 VALUES (new.id, new.properties);

--
-- DML
--

INSERT INTO g.v VALUES ('1234.56', NULL);
UPDATE g.v SET properties='{"update":"impossible"}' WHERE id = '1234.56';
DELETE FROM g.v;

-- cleanup
REVOKE ALL ON DATABASE regression FROM tmp;
DROP GRAPH t CASCADE;
DROP ROLE tmp;
DROP GRAPH g CASCADE;
