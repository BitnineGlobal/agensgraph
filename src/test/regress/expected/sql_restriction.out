--
-- SQL Restriction with Graph objects
--
-- prepare
CREATE GRAPH g;
CREATE ROLE tmp;
--
-- ALTER SCHEMA
--
ALTER SCHEMA g RENAME TO s;
ERROR:  cannot rename schema "g"
HINT:  Use RENAME GRAPH instead
ALTER SCHEMA g OWNER TO tmp;
ERROR:  cannot alter schema "g"
HINT:  Use ALTER GRAPH instead
--
-- CREATE TABLE
--
CREATE TABLE g.t (i int);
ERROR:  cannot create table in graph schema
CREATE TABLE t (i int) INHERITS (g.ag_vertex);
ERROR:  invalid parent, table cannot inherit label
--
-- ALTER TABLE
--
GRANT ALL ON DATABASE regression TO tmp;
SET ROLE tmp;
CREATE GRAPH t;
SET graph_path = t;
CREATE VLABEL v;
ALTER TABLE t.v ADD COLUMN tmp int;
ERROR:  only superuser can ALTER TABLE on graph label
RESET ROLE;
SET graph_path = g;
CREATE VLABEL v;
ALTER TABLE g.v RENAME TO e;
ERROR:  cannot rename table "v"
HINT:  Use rename label instead
--
-- TRUNCATE TABLE
--
TRUNCATE TABLE g.v;
ERROR:  cannot truncate label in graph schema
--
-- TRIGGER
--
CREATE TRIGGER tt AFTER INSERT ON g.v
FOR EACH STATEMENT EXECUTE PROCEDURE ff();
ERROR:  cannot create trigger on graph label
--
-- RULE
--
CREATE VLABEL v2;
CREATE RULE rr AS ON INSERT TO g.v DO INSTEAD
	INSERT INTO g.v2 VALUES (new.id, new.properties);
ERROR:  cannot create rule on graph label
--
-- DML
--
INSERT INTO g.v VALUES ('1234.56', NULL);
ERROR:  DML query to graph objects is not allowed
UPDATE g.v SET properties='{"update":"impossible"}' WHERE id = '1234.56';
ERROR:  DML query to graph objects is not allowed
DELETE FROM g.v;
ERROR:  DML query to graph objects is not allowed
-- cleanup
REVOKE ALL ON DATABASE regression FROM tmp;
DROP GRAPH t CASCADE;
NOTICE:  drop cascades to 4 other objects
DETAIL:  drop cascades to sequence t.ag_label_seq
drop cascades to vlabel ag_vertex
drop cascades to elabel ag_edge
drop cascades to vlabel v
DROP ROLE tmp;
DROP GRAPH g CASCADE;
NOTICE:  drop cascades to 5 other objects
DETAIL:  drop cascades to sequence g.ag_label_seq
drop cascades to vlabel ag_vertex
drop cascades to elabel ag_edge
drop cascades to vlabel v
drop cascades to vlabel v2
